### PR TITLE
feat: Create a `file` command containing `cat` and `chmod`

### PR DESCRIPTION
**Pull request by fowles**
- rearrange the files involved to be more clear about structure

Implements https://github.com/martinvonz/jj/issues/3812

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [x] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

File: cli/src/commands/file.rs [hunk 1]
```diff
@@ -14,8 +14,7 @@
 
 use crate::cli_util::CommandHelper;
 use crate::command_error::CommandError;
-use crate::commands::file_cat;
-use crate::commands::file_chmod;
+use crate::commands::{file_cat, file_chmod};
```
[the remark this answers is not in the document]
Comment (fowles) [reply cli/src/commands/file.rs]: I just squashed everything into one commit

File: cli/src/commands/file.rs [hunk 2]
```diff
@@ -0,0 +1,38 @@
+// Copyright 2020-2023 The Jujutsu Authors
```
[the remark this answers is not in the document]
Comment (fowles) [reply cli/src/commands/file.rs:1]: done

File: CHANGELOG.md [hunk 3]
```diff
@@ -21,6 +21,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Deprecations
 
+* `jj file print` (aka `jj file cat`) replaces `jj cat`.
+
+* `jj file chmod` replaces `jj chmod`.
```
[the remark this answers is not in the document]
Comment (fowles) [reply CHANGELOG.md:26]: merged all together

File: cli/testing/fake-diff-editor.rs [hunk 4]
```diff
@@ -84,7 +84,7 @@ fn main() {
                     exit(1)
                 }
             }
-            ["print", message] => {
+            ["file", "print", message] => {
```
Comment (martinvonz) [inline cli/testing/fake-diff-editor.rs:87]: This is an argument for the fake diff-editor. It should not have to match `jj file print`.

File: cli/src/commands/file.rs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+use crate::cli_util::CommandHelper;
+use crate::command_error::CommandError;
+use crate::commands::{file_chmod, file_print};
+use crate::ui::Ui;
+
+/// File operations.
+#[derive(clap::Subcommand, Clone, Debug)]
+pub enum FileCommand {
+    #[command(visible_alias("cat"))]
```
Comment (martinvonz) [inline cli/src/commands/file.rs:23]: We probably don't need this alias? At least we don't need it for backwards compatibility.

**Review by martinvonz — APPROVED** [review]
This seems like something people might have opinions about. I asked for feedback about it on Discord a day or two ago. I don't think there were any serious concerns. Two possible concerns I could think of was that these commands get a bit longer, and that it's unclear which command should be considered "file" commands.

We discussed a bit which commands should be under `jj file` and it seems like it would be `jj file print`, `jj file chmod`, and `jj file list` (the current `jj files`) for now (I know you have already prepared `jj file list`), and then at least `jj file move/mv/copy/cp` later.

No one seemed worried about the longer command names (not worried enough to mention it anyway).

Thanks!

**Comment by joyously** [comment]
> We discussed a bit which commands should be under `jj file` 

Did anyone mention `remove` or `untrack` functionality?
At first glance, the commands mentioned seemed like the argument would be a file, but when I thought again, it seemed to make sense that they could be invoked on a revset intending a file oriented output. This makes me wonder if there would be `jj file diff` or is that already `interdiff`?

File: cli/src/commands/file_chmod.rs
Comment (yuja) [inline cli/src/commands/file_chmod.rs:1]: nit: should we merge `file_*.rs` into `file.rs`? or create nested `file/{mod, chmod, print}.rs`?

(My vote is to merge them into a single `file.rs` as it's less likely that file subcommands will grow much bigger, and it's easier to rename e.g. `print` to `show`.)

File: cli/src/commands/mod.rs [hunk 6]
```diff
@@ -77,11 +78,12 @@ enum Command {
     Bench(bench::BenchCommand),
     #[command(subcommand)]
     Branch(branch::BranchCommand),
-    #[command(alias = "print")]
-    Cat(cat::CatArgs),
+    #[command(alias = "print", hide = true)]
```
Comment (yuja) [inline cli/src/commands/mod.rs:81]: nit: top-level `jj print` shouldn't be allowed?

Comment (yuja) [reply cli/src/commands/mod.rs:81]: Doh, never mind. It was existing alias.

File: cli/src/commands/file_chmod.rs
Comment (fowles) [reply cli/src/commands/file_chmod.rs:1]: I strongly prefer keeping files targeted to a single thing and not turning `file.rs` into an ever growining thing.  I could be convinced to do a subdirectory quite easily though

File: cli/src/commands/file.rs (hunk 5 above)
Comment (fowles) [reply cli/src/commands/file.rs:23]: it isn't needed for back compat, but I do think it is common enough in the linux world that we should keep it for convenience

File: cli/testing/fake-diff-editor.rs (hunk 4 above)
Comment (fowles) [reply cli/testing/fake-diff-editor.rs:87]: Done

File: cli/src/commands/file_chmod.rs
Comment (fowles) [reply cli/src/commands/file_chmod.rs:1]: went with a subdirectory

**Comment by martinvonz** [comment]
> Did anyone mention `remove` or `untrack` functionality?

Makes sense.

> This makes me wonder if there would be `jj file diff` or is that already `interdiff`?

I'm not sure where to draw the line for what goes under `jj file`. `interdiff` is quite different, though; it shows the diff between the diffs in two commits (by temporarily rebasing the first commit so it's a sibling of the second commit). It's useful for showing how a commit has evolved over time (`jj obslog -p` shows interdiffs).

File: cli/src/commands/file.rs (hunk 5 above)
Comment (yuja) [reply cli/src/commands/file.rs]: For Unix folks, I think the top-level `jj cat` alias is more important. I personally don't think Unixy command name in namespaced command would improve discoverability.

Comment (fowles) [reply cli/src/commands/file.rs]: removed for now, we can add it back later

**Pull request merged**